### DISCUSSION TITLE
Dropzone field Cloudflare images

### DISCUSSION THREAD

**Discussion by Tuoww** [Laravel-Backpack/community-forum#846] [Q&A (Help)]
Hi quick question,

I am trying to implement the use of Cloudflare Images into my application, I am using backpack with the dropzone field as the image upload mechanism. Now I fixed the upload by using an old package to create a laravel filesystem, but because the getting of the images is a bit different from what is normal in backpack I can't make that work right now. It uses a link with a unique id for every image which you get returned when uploading the image. But default backpack it uses the image filename with 4 characters appended to it.

I was wondering if you guys have ever used Cloudflare Images and/or maybe have tips on how to fix this problem.
Thanks in advance

**Comment by pxpm** [comment]
Hey @Tuoww isn't Cloudflare Images similar/compatible with s3 ? 

We will need a bit more info here, and maybe some debug information would also help. 

Run `php artisan backpack:version` and let me know what software version you are using.

It would also help to know your dropzone field definition, your disk definition and any customization your are doing to uploads. 

Uploaders use the Laravel `Filesystem` under the hood, so it should upload and retrieve the files seamless. We may have a bug, but I have no information on how to proceed here. 

Let me know more details so that I can try to help you. 🙏 

Cheers

**Reply by Tuoww** [reply]
I'll check tomorrow if it is compatible with the S3 filesystem.
https://developers.cloudflare.com/images/upload-images/upload-url/ this is the Cloudflare Images API documentation.
https://github.com/Tuoww/laravel-filesystem-cloudflare-images this is a small package i used for a laravel filesystem for cloudflare images.
`CRUD::column('photos')->type('dropzone')->withFiles(['disk' => 'cloudflare-images']);` this is the definition of the upload field. Uploading goes fine, only problem I have is when I try show the updateOperation or the listOperation I get the error 'This driver does not support retrieving URLs.'.

My main question was just if you guys have had any experience with CloudFlare images, ill update and answer to your questions tomorrow!
Thanks again

**Reply by pxpm** [reply]
I don't have specifically with Cloudflare Images, but I suspect it may be compatible with S3 so I may be able to give it a test tomorrow on an s3 compatible buck.

Come back tomorrow with your info, and I will have a look at the package you mentioned to check if it's s3 compatible. If it is, we may have an issue in dropzone. 🤷 

Just to take it out of my mind, is `cloudflare-images` your `cloud` disk in `config/filesystems.php` ? Can you try setting `temporaryUrl => true` in your uploader configuration ? 

Cheers

**Reply by Tuoww** [reply]
When trying to use my cloudflare upload endpoint:

I tried to use the s3 thing from filesystems.php like you said unfortunately that does not work because of with Cloudflare Images, because it is missing a bucket and a region and those cannot be null. I get that error message only when on the setuplistoperation. When I try to create a product with a new image in setupCreateOperation with the S3 filesystem I get this error in the logs: `[2024-02-22 13:26:26] local.ERROR: Unable to move file from backpack/temp/png-transparency-demonstration-1-qyZ1.png to png-transparency-demonstration-1-qyZ1.png`. I also tried the temporary urls but they didnt do anything.

So when i use the custom filesystem thing it does this call:
`curl --request POST \
--url https://api.cloudflare.com/client/v4/accounts/<ACCOUNT_ID>/images/v1 \
--header 'Authorization: Bearer <API_TOKEN>' \
--form 'url=https://[user:password@]example.com/<PATH_TO_IMAGE>' \
--form 'metadata={"key":"value"}' \
--form 'requireSignedURLs=false'`

Then this is the response you get if get get a successfull upload:
`{
    "result": {
        "id": "2cdc28f0-017a-49c4-9ed7-87056c83901",
        "filename": "image.jpeg",
        "metadata": {
            "key": "value"
        },
        "uploaded": "2022-01-31T16:39:28.458Z",
        "requireSignedURLs": false,
        "variants": [
            "https://imagedelivery.net/Vi7wi5KSItxGFsWRG2Us6Q/2cdc28f0-017a-49c4-9ed7-87056c83901/public",
            "https://imagedelivery.net/Vi7wi5KSItxGFsWRG2Us6Q/2cdc28f0-017a-49c4-9ed7-87056c83901/thumbnail"
        ]
    },
    "success": true,
    "errors": [],
    "messages": []
}`

To make this work in the basic form I need to save the `"id": "2cdc28f0-017a-49c4-9ed7-87056c83901"` (it now saves the filename) so I can get the image through this:
`curl -H 'Authorization: Bearer <api_token>' https://api.cloudflare.com/client/v4/accounts/a<ACCOUNT_ID>/images/v1/2cdc28f0-017a-49c4-9ed7-87056c83901` 

Thanks for your help and tips though they are appreciated!

**Reply by Tuoww** [reply]
Any idea on how to accomplish this?

**Reply by pxpm** [reply]
Hey @Tuoww sorry for my absence here.

I have never worked with that, and was not able to do any kind of progress on this issue.
So what I suggest is that you go along with a custom uploader for now or even don't use our uploaders at all.

If you still want to use "some" of the uploaders functionality, you can set the `uploader` inside your `withFiles` configuration to use your own class, and write your save/retrieve functions. https://backpackforlaravel.com/docs/6.x/crud-uploaders#configuring-the-uploaders

I am sorry I can't be much more help here, if someday I stumble on something similar I will remember to check it here but for now I have nothing that would help you. 😞 

Let me know if you find something that would help others in the same situation, maybe make a `cloudflare-uploader` package that you and other could use 🤷 

Cheers

**Reply by Tuoww** [reply]
Thanks for your suggestion, I'll look into the uploader is there an example or some more documentation for the UploaderInterface? Does this like overwrite the filesystem upload mechanism for example? If so I could use it.

**Reply by pxpm** [reply]
The uploaders are just "events", let's say. They don't overwrite nothing, they are just called at some points in your application life-cycle, do they business and move on. Those points are: `saving`, `retrieved` and if you configured `deleted`. 

Have an example of our image uploader https://github.com/Laravel-Backpack/CRUD/blob/main/src/app/Library/Uploaders/SingleBase64Image.php

There is not much docs, but you can overwrite pretty much all the uploader methods to suite your needs. 

Let me just alert you to the fact that there is a PR in progress with major changes to Uploaders architecture, so overwritting it now may not be the best solution. https://github.com/Laravel-Backpack/CRUD/pull/5440 .

Maybe just overwrite the save/update operations in your controller, and deal with it manually without relying on uploaders for now. https://backpackforlaravel.com/docs/6.x/crud-operation-update#override-the-update-method

The rewrite will also make the uploaders more extensible (at the moment they are a little bit locked), so it may even help you in your quest to develop a CloudFlare Image uploader. 

Cheers

**Reply by Tuoww** [reply]
Oh thanks for your tips, I was able to create the upload and put the cloudflare id which i got back in the db. I am using the dropzone version of the upload, but when trying to edit, the photos keep getting prepended with `http://localhost:8000/storage/` and after this is the correct url. How can I make sure that it does not show the localhost part but only my url?
